### DISCUSSION TITLE
2022/10/03/bernoulliderivations

### DISCUSSION THREAD

**Discussion by giscus** [horenbergerb/horenbergerb.github.io#16] [Announcements]
# 2022/10/03/bernoulliderivations

What is Bernoulli Diffusion? What is Bernoulli Diffusion?

https://horenbergerb.github.io/2022/10/03/bernoulliderivations.html

**Comment by PengjuYan** [comment]
Hi Beau,

Thank you for sharing this out. We found your article here when we looked for derivations of diffusion process under Bernoulli distributions. This helped us a lot!

I found there was a mistake in your previous code in the posterior calculation (`(1 - x_t) * (1.5*beta_t)`) and I'm happy that you've revised it accordingly (`(1 - x_t) * (0.5*beta_t)`).

I suggest two paper here for more formal derivations (especially about the normalization term):

1. Argmax Flows and Multinomial Diffusion: Learning Categorical Distributions
2. Structured Denoising Diffusion Models in Discrete State-Spaces

Lastly, there are still some minor typos or mistakes here:

1. $H_q(X^T)$ should be $H_p(X^T)$,
2. $\epsilon_{t}+\epsilon_{t+1}-\epsilon_{t}\epsilon_{t+1}$ should be $\epsilon_{t}+\epsilon_{t+1}-2\epsilon_{t}\epsilon_{t+1}$, and
3. $(1-c)\log c$ should be $(1-c)\log(1-c)$

Before I came across the two paper above, your blog helped me a lot by the 'flipping bits' way of thinking. Thanks again!

-Pengju

**Reply by PengjuYan** [reply]
1. Paper 1 calculates the normalization term by summing up the nominator vector values, which is one way;
2. Paper 2 calculates it straightforwardly just as you did recently.

Both are equivalent mathematically. But to be honest, I don't know which one is better considering the back-propagation (autograd) efficiency (forward propagation is the same).

**Reply by horenbergerb** [reply]
Thanks, I'm really happy to hear someone found this helpful!

Thanks for the paper suggestions! I noticed those papers recently while researching diffusion with categorical distributions, but I didn't notice that they discuss the normalization term. I will look into that!

And I'm working on your corrections now, thanks for pointing them out! I've fixed 1) and 3), but I have to review 2) a little more before I push them all
